### PR TITLE
Rename IndexValue -> Pairs, print it better

### PR DESCRIPTION
**Pull request by Keno**
As discussed in #25711

Before:
```
julia> f(;kwargs...) = kwargs
f (generic function with 1 method)

julia> f(;a = 1, b = 2)
Base.Iterators.IndexValue{Symbol,Int64,Tuple{Symbol,Symbol},NamedTuple{(:a, :b),Tuple{Int64,Int64}}} with 2 entries:
  :a => 1
  :b => 2
```

After:
```
julia> f(;a = 1, b = 2)
pairs((a = 1, b = 2)) with 2 entries:
  :a => 1
  :b => 2
```

### CODE REVIEW

File: base/essentials.jl [hunk 1]
```diff
@@ -765,3 +765,7 @@ Indicate whether `x` is [`missing`](@ref).
 """
 ismissing(::Any) = false
 ismissing(::Missing) = true
+
+# forward declarations for use by iterators.jl (since parts of that file
+# are used by the compiler)
+function showarg end
```
Comment (vtjnash) [inline base/essentials.jl]: editor forgetting to add newlines?

File: base/iterators.jl [hunk 2]
```diff
@@ -152,15 +152,23 @@ end
...
+    Pairs(data::A, itr::I) where {A, I} = new{eltype(I), eltype(A), I, A}(data, itr)
+end
+
+function Base.showarg(io::IO, r::Pairs, toplevel)
+    if try typeof(pairs(r.data)) == typeof(r); catch; false; end
+        print(io, "pairs(", r.data, ")")
+    else
+        print(io, "Iterators.Pairs(",r.data, ", ", r.itr,")")
+    end
 end
```
Comment (vtjnash) [inline base/iterators.jl]: I think we usually put these definitions in show.jl to avoid the ordering issue. Also, missing a few spaces after the commas

File: base/iterators.jl [hunk 3]
```diff
@@ -152,15 +152,23 @@ end
...
-struct IndexValue{K, V, I, A} <: AbstractDict{K, V}
+struct Pairs{K, V, I, A} <: AbstractDict{K, V}
     data::A
     itr::I
-    IndexValue(data::A, itr::I) where {A, I} = new{eltype(I), eltype(A), I, A}(data, itr)
+    Pairs(data::A, itr::I) where {A, I} = new{eltype(I), eltype(A), I, A}(data, itr)
+end
+
+function Base.showarg(io::IO, r::Pairs, toplevel)
+    if try typeof(pairs(r.data)) == typeof(r); catch; false; end
```
Comment (vtjnash) [inline base/iterators.jl]: Why is there a try/catch trying to hide on this line?

**Review by vtjnash — COMMENTED** [review]
Should we perhaps calls this a `PairsView`?

File: base/iterators.jl (hunk 3 above)
Comment (vtjnash) [reply base/iterators.jl]: This seems like a bad way to handle this.

Comment (Keno) [reply base/iterators.jl]: Seemed like `pairs` might throw?

File: base/iterators.jl (hunk 2 above)
Comment (Keno) [reply base/iterators.jl]: Can do.

File: base/iterators.jl (hunk 3 above)
Comment (Keno) [reply base/iterators.jl]: Well, this is an interactive-use function, so I didn't think it was too-terrible. The query is basically "Is this the type I'd get if I called `pairs` on this thing". Seemed like the most straightforward way to express that. We could also do something with `return_types`, but not sure that's better.

**Review by JeffBezanson — APPROVED** [review]

File: base/iterators.jl (hunk 3 above)
Comment (vtjnash) [reply base/iterators.jl]: I had a great reply typed out, then clicked the wrong button. C'est la vie, I suppose – second draft is always better, right?

I looked at how the other `showarg` functions are implemented, and noticed / realized that none of them care about whether the printed syntax is precise, just whether it is useful. That is to say, they only care if its a reasonable interpretation. None of them care to check if that interpretation is precise.

In conclusion, the implementation should simply be:
```julia
print(io, "pairs(")
showarg(io, r.data, false)
print(io, ")")
```

Comment (vtjnash) [reply base/iterators.jl]: And also define:
```julia
showarg(io, nt::NamedTuple, toplevel) = print(io, toplevel ? typeof(nt) : "::NamedTuple")
showarg(io, tt::Tuple, toplevel) = print(io, toplevel ? typeof(tt) : "::Tuple")
```

Comment (Keno) [reply base/iterators.jl]: > In conclusion, the implementation should simply be:

Well, The `Pairs` iterator allows different index sets, so e.g. `pairs(IndexCartesian(), A)` should not print as `pairs(A)`.

Comment (vtjnash) [reply base/iterators.jl]: I think the same could be said of the other `showarg` functions. Mostly, they'll be distinguished by the printed data. But we can just extend this to handle that case also.

Comment (Keno) [reply base/iterators.jl]: I don't think so. If we print a simple `pairs(` line, we should make sure it's correct. `We wouldn't print ReshapedArray as `reshape(::Array, rand(Int), rand(Int))`.
